qiskit/extensions/quantum_initializer/_initializer.py

```diff
@@ -75,6 +75,9 @@ def __init__(self, param, arg, circ=None):
         # invert the circuit to create the desired vector from zero (assuming
         # the qubits are in the zero state)
         self.inverse()
+        # do not set the inverse flag, as this is the actual initialize gate
+        # we just used inverse() as a method to obtain it
+        self.inverse_flag = False
```

Comment:
hmmm.. what about having a boolean argument to `self.inverse()` so this flag is always controlled  by the parent class? Something like:
```python
(_compositegate.py)
def inverse(initializing_gate=False):
    """Invert this gate."""
        self.data = [gate.inverse() for gate in reversed(self.data)]
        self.inverse_flag = False if initializing_gate else not self.inverse_flag
        return self
```
Object state flags are evil, so it's always good to keep them under control

Reply:
But `CompositeGate` is a fundamental class in QISKit, which captures the ability to compose multiple gates into bigger gates. The initializer gate is just one use case of it. I feel like a fundamental class should not be changed every time it is used. Also there could be other implementations of the initializer gate which don't have this inversion problem. It is specific to this implementation, thus should be managed there.

Reply:
> I feel like a fundamental class should not be changed every time it is used
Not sure I understood this, we are changing the instance state of the initializer gate.
> Also there could be other implementations of the initializer gate which don't have this inversion problem
That's why `inverse()` has a default argument value of `False`. You can pass a True just whenever you want and the code is very explicit about what's happening and restricts the usage of this flag to the parent class where is only used btw. Otherwise one just see a flag that is changing a state... a state that is only used in the parent class and only changes the behavior of this parent class... but is being changed in the child class, it feels confusing to me.
We can change the `intializing_gate` argument name to something more generic if doesn't make sense though.
Anyway, not blocking this PR because of this or course :) ... just a suggestion, I don't really like instance flags (even more on base classes).